blockchain/upgrade.go

```diff
@@ -1789,6 +1789,398 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+			for i := uint32(0); i < uint32(numBitmapBytes); i++ {
+				unspentBits := oldSerialized[offset]
+				for j := uint32(0); j < 8; j++ {
+					if unspentBits&0x01 != 0 {
+						// The first 2 outputs are encoded via the
+						// header code, so adjust the output number
+						// accordingly.
+						outputNum := 2 + i*8 + j
+						outputIndexes = append(outputIndexes, outputNum)
+					}
+					unspentBits >>= 1
+				}
+				offset++
+			}
+
+			// Create a map to hold all of the converted outputs for the entry.
+			type convertedOut struct {
+				compressedAmount uint64
+				compressedScript []byte
+				scriptVersion    uint64
+			}
+			outputs := make(map[uint32]*convertedOut)
+
+			// Decode and add all of the outputs.
+			for _, outputIndex := range outputIndexes {
+				// Deserialize the compressed amount and ensure there are bytes
+				// remaining for the compressed script.
+				compressedAmount, bytesRead := deserializeVLQ(oldSerialized[offset:])
+				if bytesRead == 0 {
+					return errDeserialize("unexpected end of data during decoding " +
+						"(compressed amount)")
+				}
+				offset += bytesRead
+
+				// Decode the script version.
+				scriptVersion, bytesRead := deserializeVLQ(oldSerialized[offset:])
+				if bytesRead == 0 {
+					return errDeserialize("unexpected end of data during decoding " +
+						"(script version)")
+				}
+				offset += bytesRead
+
+				// Decode the compressed script size and ensure there are enough bytes
+				// left in the slice for it.
+				const compressionVersion = 1
+				scriptSize := decodeCompressedScriptSize(oldSerialized[offset:],
+					compressionVersion)
+				// Note: scriptSize == 0 is OK (an empty compressed script is valid)
+				if scriptSize < 0 {
+					return errDeserialize("negative script size")
+				}
+				if len(oldSerialized[offset:]) < scriptSize {
+					return errDeserialize(fmt.Sprintf("unexpected end of "+
+						"data after script size (got %v, need %v)",
+						len(oldSerialized[offset:]), scriptSize))
+				}
+				compressedScript := make([]byte, scriptSize)
+				copy(compressedScript, oldSerialized[offset:offset+scriptSize])
+				offset += scriptSize
+
+				// Create a converted utxo entry with the details deserialized above.
+				outputs[outputIndex] = &convertedOut{
+					compressedAmount: compressedAmount,
+					compressedScript: compressedScript,
+					scriptVersion:    scriptVersion,
+				}
+			}
+
+			// Copy the stake extra data if this was a ticket.
+			var ticketMinOuts []byte
+			if txType == stake.TxTypeSStx {
+				ticketMinOuts = make([]byte, len(oldSerialized[offset:]))
+				copy(ticketMinOuts, oldSerialized[offset:])
+			}
+
+			// Create a V3 utxo entries with the details deserialized above.
+			for outputIndex, output := range outputs {
+				// Encode the V3 utxo flags.  The bit representation is:
+				//   bit  0     - containing transaction is a coinbase
+				//   bit  1     - containing transaction has an expiry
+				//   bits 2-5   - transaction type
```

Comment:
That is a fair point, though this is now reserving 4 bits for the tx types, so we won't run out of room until we need more than 16 tx types.  I don't think we will be approaching that any time soon, so I'm not sure that it is worth the effort of moving (not that it would be difficult, especially after the stxo migration is ready, as you mentioned).